### ISSUE TITLE
BTreeSet causes UB my having (partially) dangling shared ref

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#54957]
The following code causes UB (found by a patched miri enforcing validity invariants):
```rust
    let mut b = std::collections::BTreeSet::new();
    b.insert(42);
```
What happens is that `BTreeSet::new` [sets the root node](https://github.com/rust-lang/rust/blob/4623d4889373702428251967d9bf6ff6d677ec1a/src/liballoc/collections/btree/map.rs#L535) to [a pointer to](https://github.com/rust-lang/rust/blob/4623d4889373702428251967d9bf6ff6d677ec1a/src/liballoc/collections/btree/node.rs#L197) a [static `EMPTY_ROOT_NODE` shared for this purpose by all B-trees](https://github.com/rust-lang/rust/blob/4623d4889373702428251967d9bf6ff6d677ec1a/src/liballoc/collections/btree/node.rs#L116) that has type `LeafNode<(), ()>`. That step is okay.

However, in the next step, `insert` calls [`ensure_root_is_owned`](https://github.com/rust-lang/rust/blob/4623d4889373702428251967d9bf6ff6d677ec1a/src/liballoc/collections/btree/map.rs#L1174) which indirectly [turns this ptr](https://github.com/rust-lang/rust/blob/4623d4889373702428251967d9bf6ff6d677ec1a/src/liballoc/collections/btree/node.rs#L388) into `&LeafNode<i32, ()>`. The trouble is that `LeafNode<i32, ()>` is bigger than `LeafNode<(), ()>`, so this reference is partially dangling: Its beginning points to allocated memory (the static `EMPTY_ROOT_NODE`), but that allocated memory ends "too early". The invariant that `&LeafNode<i32, ()>` be dereferencable *for `mem::size_of::<LeafNode<i32, ()>>()` bytes* is violated. Since we are passing `dereferencable(n)` to LLVM [with `n` set to the full size of the struct](https://play.rust-lang.org/?gist=8c9fecbcecd4896eaa7d2a97686714da&version=nightly&mode=debug&edition=2015), this means we have UB not just according to our own validity rules, but according to LLVM's rules.

**Comment by aturon** [comment]
cc @Gankro

**Comment by Gankro** [comment]
Good catch, and very interesting. I wasn't aware that this was a thing. When was this semantic decided/documented? 

This can be easily fixed by hoisting the `ptr::eq(self, &EMPTY_ROOT_NODE as *const _ as *const _)` into `is_shared_root`.

**Comment by Gankro** [comment]
Also: what value does that metadata have? Allowing speculative loads?

**Comment by RalfJung** [comment]
@nikomatsakis points out there might also be alignment problems. @eddyb are there any types that have alignment bigger than pointers?

**Comment by nikomatsakis** [comment]
One possible fix -- but it'd be a big lang change -- is supporting generic statics. Of course that's not a "short term" usable thing.

**Comment by RalfJung** [comment]
Another thing @nikomatsakis suggested is to essentially use `&LeafNode<(), ()>` throughout the read-only operations *until* we determined that the node actually is non-empty. That should also work.

**Comment by Gankro** [comment]
Deferring working on this for a bit while some other stuff lands, some notes:

the core of the right fix is to add something like this code:

```rust
    fn as_key_slice(&self) -> &'a [K] {
        // When taking a pointer to the keys, if our key has a stricter
        // alignment requirement than the shared root does, then the pointer
        // would be out of bounds, which LLVM assumes will not happen. If the
        // alignment is more strict, we need to make an empty slice that doesn't
        // use an out of bounds pointer.
        if mem::align_of::<K>() > mem::align_of::<LeafPrefix>() && self.is_shared_root() {
            &[]
        } else {
            // Here either it's not the root, or the alignment is less strict,
            // in which case the keys pointer will point "one-past-the-end" of
            // the node, which is allowed by LLVM.
            unsafe {
                slice::from_raw_parts(
                    &(*self.node as *const LeafNode<K, V>).keys as *const K,
                    self.len()
                )
            }
        }
    }
```

to `btree::node::Root`, and introduce an untyped header/prefix as I discussed above. The annoying part is going through all the code and making perfectly sure nothing casts to a Leaf prematurely.

**Comment by RalfJung** [comment]
> the core of the right fix is to add something like this code:

I am not sure why comparing the alignments is enough to guarantee that things are in-bounds. And anyway wouldn't it be much simpler to just return `&[]` whenever `self.len() == 0`? Determining the `len` should be possible even for the "untyped header".

**Comment by Gankro** [comment]
The idea is we're trying to avoid *any* additional checks.

Alignment differences determine padding, which in turn determines if the keys would be placed within the bounds of the LeafPrefix's size.

**Comment by RalfJung** [comment]
Oh I see, the alignment checks would only involve constants and hence not be present at run-time... `self.is_shared_root()` would still be a run-time check though?

**Comment by Gankro** [comment]
Yeah it compares a dynamic pointer to a static value, but it's &&'d, so under reasonable conditions it will never run.

**Comment by Shnatsel** [comment]
Is there a way to observe some kind of memory problem in practice because of this issue, such as an out-of-bounds read?

I'm asking because it would be an interesting test for https://github.com/blt/bughunt-rust - it would be very useful to see whether the testing setup detects this issue or not in order to gauge how it performs in practice.

**Comment by gnzlbg** [comment]
> are there any types that have alignment bigger than pointers?

I'm not sure I follow whether this is relevant, but this is the case for:

* `repr(align(N))` types where `N` is larger than `align_of::<*const T>()` (e.g. if you want to efficiently save the state of all registers on x86 using something like [`_xsave`](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_xsave), you'll notice that your `Context` must be aligned to a 64-byte boundary), and
* `repr(simd)` types, which in general have alignment requirements much larger than pointers (up to 32-byte alignment).

**Comment by RalfJung** [comment]
@Gankro I have a fix for the out-of-bounds problem... but there's another one. ;)  Consider `into_slices_mut`:
```rust
    fn into_val_slice_mut(mut self) -> &'a mut [V] {
        debug_assert!(!self.is_shared_root());
        unsafe {
            slice::from_raw_parts_mut(
                self.as_leaf_mut().vals.as_mut_ptr() as *mut V,
                self.len()
            )
        }
    }

    fn into_slices_mut(self) -> (&'a mut [K], &'a mut [V]) {
        let k = unsafe { ptr::read(&self) };
        (k.into_key_slice_mut(), self.into_val_slice_mut())
    }
```
I find the `ptr::read` quite fascinating, this is duplicating a non-`Copy` value. :D

But anyway, this is a blatant violation of mutable references being unique.  Namely, when we call `into_val_slice_mut`, we call `as_leaf_mut()`, meaning we assert unique access to the entire leaf -- including the keys!  Of course, this invalidates the mutable reference returned from `into_key_slice_mut`.

Cases like this and the discussions we have been having around two-phase borrows make me wonder whether it is indeed the right call to consider creating a mutable reference as being roughly equivalent to a write access.  The trouble is, if you *don't* do that, then when do you even assert uniqueness of mutable references?  Only when they are written to?  That can't be it, we couldn't say `noalias` to LLVM with that. Passing a mutable reference to a function must also be something like a write. But then it seems really odd to special-case function calls like that, considering in particular that we want to solve https://github.com/rust-lang/rust/issues/16515 some day.

**Comment by Gankro** [comment]
The only reason we create a mutable reference here is because we lack `->`. The code is just trying to compute some pointer offsets as easily as possible :/

**Comment by RalfJung** [comment]
Ah right, using raw pointers likely fixes the problem.

@nikomatsakis @eddyb lack of auto-deref for raw pointers considered harmful...

**Comment by eddyb** [comment]
I'd rather change the deref operator from prefix to postfix than make it implicit for raw pointers.

**Comment by RalfJung** [comment]
Yeah, a postfix deref operator would be nice. Currently, [things get a bit ugly](https://github.com/RalfJung/rust/commit/8d1d9a1a58b4129f7abd7228cb644d3432ecde1a).

**Comment by glaebhoerl** [comment]
Couldn't we add some kind of method (inherently postfix) as an 80-20 solution? Or is the problem that we couldn't give it an appropriate type?

**Comment by RalfJung** [comment]
I don't think `*` is an operation that a function can perform.  `*` does not mean memory actually gets accessed, as in `&*`!

**Comment by Gankro** [comment]
I don't think postfix `*` works because `ptr*.0` is ambiguous as either "mul by .0" or "deref to get field .0", would likely need `->` or something new.

**Comment by RalfJung** [comment]
I submitted a proposed fix for this bug at https://github.com/rust-lang/rust/pull/56648
